File: Week3/Code/Vectorize1.R
#!/usr/bin/env Rscript
# Author: Eva Linehan
# Date: October 2018
# Desc: Vectorization example to compare the process time of vectors in comparison to loops.


#clear environments
rm(list=ls())

M <- matrix(runif(1000000),1000,1000)

SumALLELEMENTS <- function(M) {
  Dimensions <- dim(M)
  Tot<-0
  for (i in 1:Dimensions[1]){
    for (j in 1:Dimensions[2]){
      Tot <- Tot+ M[i,j]
    }
  }
  return(Tot)
}

## This on my computer takes about 1 sec
result <- as.numeric(system.time(SumALLELEMENTS(M)))
cat(paste("The execution time for Vectorize.1 R was", result[3], "\n"))

#a = Sys.time()
#Sys.sleep(SumALLELEMENTS(M))
#b = Sys.time()    
#print(paste0(round(as.numeric(difftime(time1 = b, time2 = a, units = "secs")), 3), " seconds"))
## While this takes about 0.01 sec
#print(system.time(sum(M)))

File: Week3/Code/Vectorize1.py
#!usr/bin/env python3
"""  Vectorization example to compare the process time of vectors in comparison to loops. """
__author__= 'Eva Linehan (eva.linehan18@imperial.ac.uk)'
__version__ = 0.01
__date__ = 'Dec 2018'
__licence__ = 'Inclass practical' 

import numpy as np
import scipy as sc
import time

M = np.random.rand(1000, 1000)
#M = sc.matrix(M)
#print(M.shape)

def SumALLELEMENTS(M):
    Dimensions = M.shape
    Tot = 0
    for i in range(1,Dimensions[0]):
        for j in range(1,Dimensions[1]):
            Tot = Tot + M[i,j]
    return(Tot)


t0 = time.time()
SumALLELEMENTS(M)
t1 = time.time()
runtime = t1-t0


print("The execution time for Vectorize.1 py was {}".format(runtime))
